### PR TITLE
fix(gnodev): timestamp issue on reload

### LINKED ISSUE #1509: gnodev time.Now() values are not preserved on reload
### Description

In gnodev, the reload operation replays all the transactions. In general this lets the realm functions restore the state. But if some realm functions use `time.Now()` then this returns the (arbitrary) time of reload, not the time of the original transaction. Therefore, reload does not actually restore the state. 
### Your environment

* macOS Ventura 13.6.3
* Go 1.20.7
* Latest gnolang/gno as of posting
* Latest commit: c6097cd83326cae0262b20b1fd7dff9e068caf75

### Steps to reproduce

* Set up the [test1 key for gnokey](https://github.com/gnolang/gno/blob/master/gno.land/cmd/gnofaucet/README.md)
* Install gnodev
* Start gnodev
* Register the test1 user, create a board and post with `CreateThread` as follows:
```
gnokey maketx call -pkgpath "gno.land/r/demo/users" -func "Register" -args "" -args "test_1" -args "Profile description" -gas-fee "10000000ugnot" -gas-wanted "2000000" -send "200000000ugnot" -broadcast -chainid tendermint_test -remote 127.0.0.1:36657 test1
gnokey maketx call -pkgpath "gno.land/r/demo/boards" -func "CreateBoard" -args "myboard" -gas-fee "1000000ugnot" -gas-wanted "5000000" -broadcast -chainid tendermint_test -remote 127.0.0.1:36657 test1
gnokey maketx call -pkgpath "gno.land/r/demo/boards" -func "CreateThread" -gas-fee 1000000ugnot -gas-wanted 5000000 -broadcast -chainid tendermint_test -args "1" -args " mytitle" -args "post at <current UTC time>" -remote "127.0.0.1:36657" test1
```
* View the board at http://127.0.0.1:8888/r/demo/boards:myboard . Note the timestamp on the message.
* Wait 5 minutes. In gnodev, press R to reload

### Expected behaviour

The boards realm [uses `time.Now()`](https://github.com/gfanton/gno/blob/6cabad4f9b09bf4a1e03b27fbc0d666467f2ced6/examples/gno.land/r/demo/boards/post.gno#L52) for the displayed post time.  Reload should restore the realm state including timestamps from `time.Now()`

### Actual behaviour

The timestamps of all message have the time of reload, erasing important information. This is because to reload gnodev resends the transactions where the block time is the current time of reload. This same timestamp is returned by every call to `time.Now()` .

(Related, the boards realm in testnet3 was built by replaying transactions, similarly to gnodev reload. The original message timestamps are erased and are set to the same value "2023-08-18 11:44am UTC". https://test3.gno.land/r/demo/boards:testboard .) 

### Proposed solution

One possible solution: The transaction processing code should have an option to specify the block time. Instead of using a high-level broadcast during replay, gnodev should call the lower-level transaction processing code and supply the correct timestamp of the original transaction. This will be used by `time.Now()`.

Here are the message times before reload:
Screenshot 2024-01-10 at 10 15 01

Here is where all timestamps are reset after reload:
Screenshot 2024-01-10 at 10 17 17

### ⚠️ Risks and impact if not fixed

A realm developer can't use gnodev with any realm that relies on `time.Now()`. For example:
* Timestamps on posted messages, or any time-based calculations like "number if messages in the last day".
* There are hacks to avoid using `time.Now()` but they introduce security risks of not having reliable timestamps from the transaction.
* The long-term technique to do realm versioning may also solve this problem by replaying transactions, and so this also needs to be addressed in this case.

### PR DESCRIPTION
**Pull request by gfanton**
- [x] depends on #2941  

resolve #1509  

This PR addresses the timestamp issue on gnodev by implementing the `MetadataTX` changes from #2941. Timestamps will now be correctly handled for `Reload` and `import`/`export`. For `Reset`, the timestamp will be updated to the current time.  

cc @zivkovicmilos @thehowl  

#### ~TODOs~ 
-  [x] test replays (I've only tested it manually)  

<details><summary>Contributors' checklist...</summary>

- [ ] Added new tests, or not needed, or not feasible
- [ ] Provided an example (e.g. screenshot) to aid review or the PR is self-explanatory
- [ ] Updated the official documentation or not needed
- [ ] No breaking changes were made, or a `BREAKING CHANGE: xxx` message was included in the description
- [ ] Added references to related issues and PRs
- [ ] Provided any useful hints for running manual tests
- [ ] Added new benchmarks to [generated graphs](https://gnoland.github.io/benchmarks), if any. More info [here](https://github.com/gnolang/gno/blob/master/.benchmarks/README.md).
</details>

### CODE REVIEW

**Review by zivkovicmilos — APPROVED** [review]
Clean 💯

**Comment by zivkovicmilos** [comment]
Codecov making the CI fail because this is on a fork 🤦‍♂️

File: contribs/gnodev/pkg/dev/node_test.go [hunk 1]
```diff
@@ -221,6 +223,118 @@ func Render(_ string) string { return str }
...
+	foopkg := generateTestingPackage(t, "gno.mod", foobarGnoMod, "foo.gno", fooFile)
+
+	// Call NewDevNode with no package should work
+	node, emitter := newTestingDevNode(t, foopkg)
+	assert.Len(t, node.ListPkgs(), 1)
+
+	// Span multiple time
+	for i := 0; i < 2; i++ {
+		// Wait a little for time shift
+		time.Sleep(time.Second)
```
[the remark this answers is not in the document]
Comment (zivkovicmilos) [reply contribs/gnodev/pkg/dev/node_test.go:288]: This is probably a major nit and maybe an over-optimization, but if you want to mock the current time, just have a callback for fetching the current time in the node, and modify it in the test to return a specific time, or something similar

[the remark this answers is not in the document]
Comment (gfanton) [reply contribs/gnodev/pkg/dev/node_test.go:288]: If we can do better, I'm in, but I'm not sure what you have in mind by mocking the current time and still having a pertinent test. 
A good compromise for me would have been to use `100*millisecond` instead, but it seems Node only takes `second` into account on load for now.
 Ultimately, if it has to stay like this, I will add a detailed comment to explain the reason for the sleep.

File: contribs/gnodev/pkg/dev/node.go [hunk 2]
```diff
@@ -173,9 +179,10 @@ func (n *Node) getBlockTransactions(blockNum uint64) ([]gnoland.TxWithMetadata,
 
 	txs := make([]gnoland.TxWithMetadata, len(b.Block.Data.Txs))
 	for i, encodedTx := range b.Block.Data.Txs {
+		// fallback on std tx
 		var tx std.Tx
 		if unmarshalErr := amino.Unmarshal(encodedTx, &tx); unmarshalErr != nil {
-			return nil, fmt.Errorf("unable to unmarshal amino tx, %w", unmarshalErr)
+			return nil, fmt.Errorf("unable to unmarshal tx: %w", err)
```
Comment (ltzmaxwell) [inline contribs/gnodev/pkg/dev/node.go:185]: err should be `unmarshalErr`?

**Pull request merged**